docs: sample feature for tutorials

NicolasMassart opened the pull request:
💬 **Status**: Request for Comment 

> [!NOTE]
> This PR is too big, on purpose at it aims at isolating the sample feature in one place.
> Initially it was not planned to merge it in main, ever. Like on extension.
> But the way I finally built it would make it fine to merge as it's totally isolated.
> Once the whole feature is reviewed and tested, given that it's totally isolated from the main app and not shipped into production, I see no issue to override the size check.

> [!IMPORTANT]
> ## 📋 Review Focus: Files Outside Sample Feature Directory
>
> While most of the sample feature code is contained within `app/features/SampleFeature/`, the following files **outside that directory** integrate the feature and require careful review:
> 
> ### 🔧 Engine Integration (Critical)
> - **`app/core/Engine/Engine.ts`** - Controller initialization
> - **`app/core/Engine/types.ts`** - Type definitions with conditional compilation
> - **`app/core/Engine/messengers/index.ts`** - Messenger configuration
> - **`app/core/Engine/utils/utils.test.ts`** - Engine test mocks
> 
> **What to check:** Ensure conditional compilation blocks (`///: BEGIN:ONLY_INCLUDE_IF(sample-feature)`) are correctly placed and the feature can be toggled on/off without breaking the build.
> 
> ### 🗂️ State Management
> - **`app/reducers/index.ts`** - Sample counter reducer integration
> - **`app/util/test/initial-root-state.ts`** - Initial state setup
> 
> **What to check:** Reducer is properly feature-flagged and doesn't affect other parts of the app.
> 
> ### 🧭 Navigation & UI
> - **`app/components/Views/Settings/index.tsx`** - Developer Options entry point
> - **`app/constants/navigation/Routes.ts`** - Route definitions
> 
> **What to check:** Navigation flows work correctly and Developer Options menu renders properly.
> 
> ### 📊 Analytics & Utilities
> - **`app/core/Analytics/MetaMetrics.events.ts`** - Sample feature analytics events
> - **`app/util/trace.ts`** - Tracing functionality
> 
> ### 🔬 Testing
> - **`e2e/selectors/Settings/DeveloperOptions.selectors.ts`** - E2E selectors
> - **`e2e/tags.js`** - E2E test tags
> - **`appwright/utils/Flows.js`** - E2E flows
> 
> **What to check:** E2E tests can locate and interact with sample feature components.
> 
> ### ⚙️ Build Configuration
> - **`metro.transform.js`** - Feature flag configuration
> 
> **What to check:** `'sample-feature'` is correctly added to `availableFeatures`.

## **Description**

This PR delivers a complete sample feature implementation demonstrating best practices for feature development in MetaMask Mobile. It includes new components for a counter, pet names management, network display, a dedicated developer settings entry point, comprehensive unit and end-to-end tests, and corresponding configuration/documentation updates.

Read the [README.md](https://github.com/MetaMask/metamask-mobile/blob/feat/4428_sample_feature/app/features/SampleFeature/README.md) in this PR branch for more info.

## Overview

The Sample Feature is a development and testing feature that demonstrates best practices and patterns for building features in MetaMask Mobile. This feature serves as a comprehensive reference implementation showcasing various aspects of the application architecture.

> [!WARNING]
> This is a sample feature for development/testing purposes only and should not be activated in production builds.

```mermaid
graph TB
    %% Developer Actor
    Dev[👨‍💻 Developer]
    
    %% Developer Use Cases
    Dev --> ActivateFeature[Activate Sample Feature with env var flag]
    Dev --> RunE2ETests[Run E2E Tests]
    Dev --> RunUnitTests[Run Unit Tests]
```

```mermaid
graph TB
    %% User Actor
    User[👤 User]
    
    %% User Use Cases
    User --> IncrementCounter[Increment Counter]
    User --> AddPetName[Add New Pet Name]
    User --> UpdatePetName[Update Pet Name]
    User --> SwitchNetwork[Switch Network]
    SwitchNetwork --> SeePetNamesEffect[See Effect on Pet Names List]
```

**Core Implementation**
- [x] New route & page setup
- [x] React component architecture
- [x] State management (Redux ToolKit)
- [x] API controller integration

**Quality, Observability & Operations**
- [x] E2E test
- [x] Error handling patterns
- [x] Feature flag configuration
  - [x] with code fence and env var
  - [x] with remote feature flag controller
- [x] Unit tests
- [ ] ~Storybook stories~ (not relevant, no new UI item created)
- [x] Analytics event tracking
- [x] Performance tracing
- [x] Documentation (code doc, readme, ...)

## **What's Implemented**

### Feature Components
- **Sample Counter**: Redux Toolkit state management demo with increment/decrement
- **Sample Pet Names**: Controller-based state management with form validation and network-specific persistent data storage
- **Sample Network Display**: Network state integration showing current network info
- **Developer Settings Entry Point**: Integration into app settings for feature access

### Technical Highlights
- **Codes Structure**: Sample Feature mirrors the `/app` folder organisation
- **Code Fencing**: Complete build-time exclusion using `///: BEGIN:ONLY_INCLUDE_IF(sample-feature)` 
- **Analytics**: MetaMetrics event tracking with typed events (counter increments, pet name operations)
- **Testing**: Comprehensive unit tests for all components and E2E tests
- **Form Validation**: Address validation, required fields, duplicate handling with confirmation dialogs
- **Error Handling**: Try-catch patterns, user-friendly error messages, validation feedback
- **Navigation**: Proper React Navigation integration with Settings → Developer Options → Sample Feature flow

Use this as a template for future feature development! Let me know if anything needs clarification.

> [!NOTE]
>  See also https://github.com/MetaMask/core/pull/5363 for the sample controllers used in this sample feature

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null

## **Related issues**

fixes https://github.com/MetaMask/MetaMask-planning/issues/4428

## **Manual testing steps**

See [Sample feature manual testing scenarios](https://github.com/MetaMask/metamask-mobile/blob/feat/4428_sample_feature/app/features/SampleFeature/e2e/sample-scenarios.feature) in this PR branch for complete testing scenarios.

1. Enable the feature: add `export INCLUDE_SAMPLE_FEATURE=true` in `.js.env` or run `INCLUDE_SAMPLE_FEATURE=true yarn start:ios` or `INCLUDE_SAMPLE_FEATURE=true yarn start:android`
2. Enable Developer Mode in app settings (as the sample feature is accessed from Developer Options setting)
3. Navigate to Settings → Developer Options
4. Tap "Navigate to Sample Feature"
5. Test counter increment
6. Test pet names creation and update
7. Switch networks and verify network-specific pet names

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

This is a completely new feature

### **After**

#### Unit Tests

<img width="400" alt="image" src="https://github.com/user-attachments/assets/75b6abf8-d3b7-477e-92d3-bb261bc21b61" />

all unit tests:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/65e234fa-d403-4fd8-acd5-2d88dc4c0fe7" />

#### E2E Tests

<img width="400" alt="image" src="https://github.com/user-attachments/assets/0b4bac7f-baae-466a-b37f-e4228eb1da42" />

<img width="400" alt="image" src="https://github.com/user-attachments/assets/cc125961-d796-40eb-86ea-2c7909aa3799" />

<img width="400" alt="image" src="https://github.com/user-attachments/assets/a8506065-50b4-4af9-a158-7b63c2f06ae0" />

https://github.com/user-attachments/assets/1b3ae94c-b7d8-44eb-8acf-81dc3b3a1ce5

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a development-only Sample Feature (counter, pet names, network display) gated by env/code-fencing, with navigation entry, controller/state integration, analytics/tracing, and comprehensive tests.
> 
> - **Sample Feature (dev-only, code-fenced)**
>   - UI: `SampleFeature` screen with `SampleCounterPane`, `SamplePetNames`, `SampleNetworkDisplay` and Dev Settings entry point.
>   - Feature flags: build-time `INCLUDE_SAMPLE_FEATURE`, local `MM_SAMPLE_FEATURE_COUNTER_ENABLED` (with remote fallback).
> - **Navigation/Settings**
>   - Adds `Routes.SAMPLE_FEATURE` and registers screen/flow in `MainNavigator` (fenced).
>   - Developer Options includes "Navigate to Sample Feature" section (fenced).
> - **Engine/State Integration**
>   - Integrates `@metamask/sample-controllers` `SamplePetnamesController`: init, messenger, types, state exposure, and state change events (all fenced).
>   - Redux: adds `sampleCounter` slice; initial test state wired.
> - **Analytics & Tracing**
>   - Adds MetaMetrics events for Sample Feature and Sentry trace names/ops (fenced).
> - **Build/Config**
>   - Metro transformer updated to include `sample-feature` when `INCLUDE_SAMPLE_FEATURE=true`.
>   - Jest/Sonar: exclude Sample Feature E2E from coverage/tests; add tsdoc tag.
>   - Env example adds flags for including feature and local counter toggle.
> - **Tests**
>   - Unit tests and snapshots for screens/hooks; E2E pages/specs to navigate and validate counter/pet names flows; settings selectors updated.
> - **Localization & Deps**
>   - Adds English strings for Sample Feature and Dev Options.
>   - Adds dependency: `@metamask/sample-controllers`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 545cb68c9ef230201749283979f43fc663d4d6e8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Cal-L commented:
I may be missing context on the purpose of the chosen sample feature UI - I would recommend that we simplify the feature and keep the functionality as isolated as possible. A few reasons for this:
- Keeping this as barebones as possible and keeping the logic completely isolated from production code will reduce the upkeep and lessen the chances of the sample breaking
- More components & more hooks != more help, I would argue to say less is more here and less code would also reduce the noise

NicolasMassart committed:
reapply all sample feature changes on freshly reset branch

Otherwise impossible to build when trying to simply update with main

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
fix unit test

NicolasMassart committed:
update test and add cursor rule for testing

NicolasMassart committed:
test(sample-feature): update test snapshots and test files for sample feature implementation

NicolasMassart committed:
chore: add commit rules documentation

NicolasMassart committed:
Refactor SampleFeature to use Redux only

NicolasMassart committed:
update doc

NicolasMassart committed:
feat: implement SamplePetnamesController using @metamask/sample-controllers

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

# Conflicts:
#	app/core/Engine/Engine.ts

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
fix unit tests

NicolasMassart committed:
format

NicolasMassart committed:
lint fix

NicolasMassart committed:
fix tests

NicolasMassart committed:
remove unwanted change

NicolasMassart committed:
fix sonar

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
rename snapshot tests

NicolasMassart committed:
update snapshots

NicolasMassart committed:
fix unit test

NicolasMassart committed:
readonly props

NicolasMassart committed:
updated snapshot

NicolasMassart committed:
fix path

NicolasMassart committed:
update to use controller propely

was still using the addressbook controller...
and added metrics

NicolasMassart committed:
update tests

NicolasMassart committed:
format

NicolasMassart committed:
move to top to prevent always conflicting

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
update styles

NicolasMassart committed:
remove unused import

NicolasMassart committed:
update styles

NicolasMassart committed:
fix controller selector

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
lint

NicolasMassart committed:
fix sonar issues

NicolasMassart committed:
e2e and doc

NicolasMassart committed:
update snapshot

NicolasMassart committed:
fix unit test command

to only run unit tests based on file names

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

# Conflicts:
#	ios/Podfile.lock

NicolasMassart committed:
code fence

NicolasMassart committed:
more e2e

NicolasMassart committed:
more e2e

NicolasMassart committed:
fix copilot feedback

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

# Conflicts:
#	app/core/Engine/Engine.ts
#	app/core/Engine/constants.ts
#	app/core/Engine/messengers/index.ts
#	app/core/Engine/types.ts
#	app/core/Engine/utils/utils.test.ts
#	app/util/logs/__snapshots__/index.test.ts.snap
#	metro.transform.js

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
update readme

updated snapshot

NicolasMassart committed:
update snapshot

NicolasMassart committed:
exclude e2e from coverage

NicolasMassart committed:
add sentry performance tracing

NicolasMassart committed:
feat(sample-feature): implement privacy-focused Sentry tracing

- Remove address and name tracking from Sentry traces for privacy protection
- Keep chainId and petNamesCount tracking (safe public/aggregate data)
- Update tests to reflect privacy-focused trace data structure
- Update README documentation to clarify privacy approach
- Maintain performance monitoring capabilities while protecting user data

This ensures user privacy is protected while still providing valuable
performance monitoring and debugging capabilities.

NicolasMassart committed:
remove sensitive info from tracking

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
fix type for engine

for this specific one, we have to use interface as asked by lint, despite consistency

NicolasMassart committed:
remote feature flag

NicolasMassart committed:
lint fix

NicolasMassart committed:
fix tsc issue

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
exclude e2e from sonar coverage check

NicolasMassart committed:
remove cursor rules

should be in another PR

NicolasMassart committed:
do not check dup on tests

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

Cal-L committed:
Merge from main

Cal-L committed:
Remove redundant fence tags

Cal-L committed:
Merge with main

Cal-L committed:
Format files

Cal-L committed:
Fix test snapshot

Cal-L committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat/4428_sample_feature

Cal-L committed:
Fix snapshots

Mrtenz reviewed (commented):
I don't understand why we're adding a bunch of dead code to actual source code of the app. Wouldn't it be better to improve the documentation instead?

NicolasMassart committed:
Merge main branch into feat/4428_sample_feature - resolve all conflicts and update to latest main

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
fix: maintain consistent Set return type in getBuildTypeFeatures function

- Fix bug where getBuildTypeFeatures would return Array instead of Set when INCLUDE_SAMPLE_FEATURE is enabled
- Change features = [...features, 'sample-feature'] to features.add('sample-feature')
- Ensures callers always receive Set type regardless of sample feature inclusion
- Prevents unexpected behavior for code that expects consistent return type

NicolasMassart committed:
Merge branch 'feat/4428_sample_feature' of github.com:MetaMask/metamask-mobile into feat/4428_sample_feature

NicolasMassart committed:
fix snapshot

NicolasMassart committed:
fix: exclude all e2e folders from Jest unit test runs

- Replace hardcoded e2e paths with glob pattern **/e2e/
- Automatically ignores any e2e folder at any level in project structure
- More maintainable and scalable than hardcoded paths
- Prevents e2e tests from being picked up by unit test runner
- E2e tests should only run with Detox framework

NicolasMassart committed:
fix unit test

NicolasMassart committed:
fix unit test

NicolasMassart committed:
fix: resolve TypeScript errors in SampleFeature e2e tests

- Fix import paths to use correct e2e framework structure
- Replace deprecated assertion methods with modern equivalents:
  - checkIfVisible -> expectElementToBeVisible
  - checkIfTextIsDisplayed -> expectTextDisplayed
- Add new RegressionSampleFeature tag for Sample Feature tests
- Remove unnecessary clear fields steps for cleaner test logic
- Maintain original test structure and functionality
- Remove unused TestHelpers imports

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

Cal-L reviewed (dismissed):
LGTM

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
fix(sample-feature): resolve TypeScript errors in test mocks and E2E imports

NicolasMassart committed:
fix(sample-feature): import controller events from package and fix conditional compilation

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
update snapshots

NicolasMassart committed:
Merge branch 'feat/4428_sample_feature' of github.com:MetaMask/metamask-mobile into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature

Gudahtt reviewed (approved):
Metro transform LGTM, approving on behalf of core-platform

NicolasMassart committed:
Merge branch 'main' into feat/4428_sample_feature
